File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        ArrayList<Integer> ans = new ArrayList<>();
        HashMap<String, Integer> dictionary = new HashMap<String, Integer>();
        int dicIdx = 1;
        for(char i='A'; i<='Z'; i++)
            dictionary.put(i+"", dicIdx++);
        
        int idx = 0;
        while (idx < msg.length()){
            int tmpIdx = 0;
            int cnt = 0;
            for (int i = idx+1; i<= msg.length(); i++){
                String tmp = msg.substring(idx, i);
                if (dictionary.containsKey(tmp)){
                    tmpIdx = dictionary.get(tmp);
                    cnt++;
                }
                else{
                    dictionary.put(tmp, dicIdx++);
                    break;
                }
            }
            idx += cnt;
            ans.add(tmpIdx);
        }
                
        int[] answer = new int[ans.size()];
        for(int i=0; i<ans.size(); i++)
            answer[i] = ans.get(i);
        
        return answer;
    }
}

File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    dictionary = {}
    dicIdx= 27

    # 사전 생성
    for i in range(1, 27):
        ch = chr(i+(ord('A')-1))
        dictionary[ch] = i
    
    idx = 0
    while idx < len(msg):
        tmpIdx = 0
        cnt = 0 # 아래의 for문에서 idx를 그대로 사용하기 위하여 횟수를 세어주는 변수 선언
        for j in range(idx+1, len(msg)+1):
            tmp = msg[idx:j]
            if tmp in dictionary:
                tmpIdx = dictionary[tmp]
                cnt += 1
            else:
                dictionary[tmp] = dicIdx
                dicIdx += 1
                break
        idx += cnt
        answer.append(tmpIdx)
        
    return answer